transports/quic/src/transport.rs

```diff
@@ -0,0 +1,707 @@
+
+//! Implementation of the [`Transport`] trait for QUIC.
+//!
+//! Combines all the objects in the other modules to implement the trait.
+
+use crate::connection::Connection;
+use crate::endpoint::ToEndpoint;
+use crate::provider::Provider;
+use crate::{endpoint, muxer::QuicMuxer, upgrade::Connecting};
+use crate::{Config, ConnectionError};
+
+use futures::channel::{mpsc, oneshot};
+use futures::future::{BoxFuture, MapErr};
+use futures::ready;
+use futures::stream::StreamExt;
+use futures::{prelude::*, stream::SelectAll};
+
+use if_watch::{IfEvent, IfWatcher};
+
+use futures::channel::oneshot::{Canceled, Receiver};
+use libp2p_core::{
+    multiaddr::{Multiaddr, Protocol},
+    transport::{ListenerId, TransportError as CoreTransportError, TransportEvent},
+    PeerId, Transport,
+};
+use quinn_proto::ConnectError;
+use rand::prelude::SliceRandom;
+use rand::thread_rng;
+use std::collections::hash_map::Entry;
+use std::collections::{HashMap, VecDeque};
+use std::marker::PhantomData;
+use std::net::IpAddr;
+use std::task::Waker;
+use std::{
+    net::SocketAddr,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+#[derive(Debug)]
+pub struct GenTransport<P> {
+    config: Config,
+    listeners: SelectAll<Listener>,
+    /// Dialer for each socket family if no matching listener exists.
+    dialer: HashMap<SocketFamily, Dialer>,
+
+    _marker: PhantomData<P>,
+}
+
+impl<P> GenTransport<P> {
+    pub fn new(config: Config) -> Self {
+        Self {
+            listeners: SelectAll::new(),
+            config,
+            dialer: HashMap::new(),
+            _marker: Default::default(),
+        }
+    }
+}
+
+/// Error that can happen on the transport.
+#[derive(Debug, thiserror::Error)]
+pub enum TransportError {
+    /// Error while trying to reach a remote.
+    #[error(transparent)]
+    Reach(#[from] quinn_proto::ConnectError),
+    /// Error after the remote has been reached.
+    #[error(transparent)]
+    Established(#[from] ConnectionError),
+
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
+
+    /// The task driving the endpoint has crashed.
+    #[error("Endpoint driver crashed")]
+    EndpointDriverCrashed,
+}
+
+impl<P: Provider> Transport for GenTransport<P> {
+    type Output = (PeerId, QuicMuxer);
+    type Error = TransportError;
+    type ListenerUpgrade = Connecting;
+    type Dial = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn listen_on(
+        &mut self,
+        addr: Multiaddr,
+    ) -> Result<ListenerId, CoreTransportError<Self::Error>> {
+        let socket_addr = multiaddr_to_socketaddr(&addr)
+            .ok_or(CoreTransportError::MultiaddrNotSupported(addr))?;
+        let listener_id = ListenerId::new();
+        let listener = Listener::new::<P>(listener_id, socket_addr, self.config.clone())?;
+        self.listeners.push(listener);
+
+        // Remove dialer endpoint so that the endpoint is dropped once the last
+        // connection that uses it is closed.
```

Comment:
Open question (to me): Do we have a test that runs this flow? i.e. dial first, then listen_on, then make sure the dialer's endpoint is cleaned up eventually.

Reply:
We don't. I think this is fine for now (no test needed here) but probably something that would be nice to have in the future.

Reply:
Added tests for this in 2eafbe6d430534ee3f0275b02209768375519ec5 and c33d17f370923f0675eee2ff9434eb28f7d35f15.